### ISSUE TITLE
Tag System

### ISSUE THREAD

**Issue by UnknownVoid** [aura-project/aura#501]
As we know all monsters in the game have tags such as not being knocked back or downed even preventing the use of certain skills on bosses.

Enemys like glas need this system in order to work properly.

**Comment by Sairii** [comment]
o.O....

I thought we had that in the race db, with exception of preventing certain skills ( is there really such a thing??)

we do have knockcount and stand. (stand: 0x01 Knockback-able, 0x02 Knockdown-able, 0x03 Both)
0x00 is no knockback and knockdown me thinks.

**Comment by UnknownVoid** [comment]
Yep. The tag system specifys tons of data :), you can control what is immune to any effect, skill, reforges and even display boss hp bar for normal monsters.

**Comment by UnknownVoid** [comment]
here are following tags that are importent
/not_down/   (Can't be knocked down)
unable_losesight/ (can't lose sight of the player)
no_delay/  (no reaction delay or ai delay, not sure...)
no_hitmotion/ (No motion on hit (knock back?)
unable_tame/ (Can't be tamed)
unable_bubble/ (Pet skill can't used Bubble on this npc (G2 or G3 Pet Skill)
not_swallow/    (Can't use swallow skill on this mob (Mimic Skill)
unable_sharpmind/ (Can't see the skill load with sharp mind)
not_freeze/ (Can't Freeze, Ice Spear Freeze Effect)
no_blowaway/ (Using the Alchemy Wind Skill will not push the enemy back)
no_stone/     (Can't turn this mob into stone (using certain skills or effects)
no_pounding/ (Can't use the fighter pounding skill on this mob)
not_raincasting/ (Can't be debufffed using rain casting alchemy skill)
no_counterattack/ (Counter Attack will have no effect when mobs attack you..) 
sketchable/ (G7-G9 Feature, Allows this monster to be sketched by the sketch skill)

There are tons more, but i just listed the most relevent ones for future content, incase anyone wants to make Alchemy skills or pet skills before there time :)

**Comment by eotetb** [comment]
Is there a special tag for field bosses? Field boss cp is supposed to be hidden from the player (as in, it just shows up as normal regardless of the cp of the player), I'm kind of wondering if that feature is part of the tag system or something separate entirely.

**Comment by exectails** [comment]
>Field boss cp is supposed to be hidden from the player

Really? That's interesting =O I just discovered recently that monsters don't have any power rating displayed if they're spawned with the NamedNpc state. I guess we know what that combination is for now.

>Is there a special tag for field bosses?

Even if there were, tags are client sided, we only have them for reference.

>Yep. The tag system specifys tons of data

What Sairii was referring to was the following, which was added long before we even really handled tags.
https://github.com/aura-project/aura/blob/master/system/db/races.txt#L15

This is based on the client's `BlowAwayEnable` and `ShovedEnable` attributes. You might notice that Glas actually doesn't have any of those tags you mentioned. His behavior is defined via the "stand" as Aura calls it, 0x01 for Glas,, making him knock *back*-, but not knock *down*able.

You are right about it not being handled properly though, and we definitely have to fix that. Remind me, you can't knock him down, but that's what we currently do, resulting in a kind of static fly back animation.  He can be knocked back though, which would be his stumble animation.

Now, *how* can you knock him back? Do normal attacks work? Smash? Windmill? Firebolt?

**Comment by exectails** [comment]
Another example of tags being ambiguous is the sharp mind immutability you mentioned. While there is the `unable_sharpmind` tag, some monsters have a natural immunity, like Wisp, which is the reason I added the `sharpMindImmune` field, instead of using the tag and hard-coding race ids.

**Comment by Sairii** [comment]
pretty sure firebolt can ........ dunno something aka shove or knockback
